src/templateengine.vala

```diff
@@ -54,117 +59,79 @@ namespace Valdo.TemplateEngine {
-                .replace_eval (template_child_path_relative, template_child_path_relative.length, 0, 0, (match_info, result) => {
-                    string variable_name = (!)match_info.fetch (1);
-
-                    if (variable_name in substitutions) {
-                        result.append (substitutions[variable_name]);
-                    } else {
-                        warning ("could not substitute `${%s}` in %s - prepend a `$` if this was intentional",
-                                 variable_name, template_child_path_relative);
-                        result.append (variable_name);
-                    }
-
-                    return false;
-                });
-
-            var project_child = project_dir.resolve_relative_path (project_child_path_relative);
-            var project_child_parentdir = project_child.get_parent ();
-
-            if (file_type == FileType.DIRECTORY) {
-                // create an empty directory
-                DirUtils.create_with_parents ((!) project_child.get_path (), 0755);
-            } else {
-                // create the parent directory of the file
-                if (project_child_parentdir != null) {
-                    DirUtils.create_with_parents ((!) (/* FIXME: non-null */ (!)project_child_parentdir).get_path (), 0755);
-                }
+            var template_file = files_list[fileinfo];
+            var relative_path = (!) template.directory.get_relative_path (template_file);
 
-                // if this is not a template file, just copy it over
-                if (!(template_child_path_relative in template.inputs)) {
-                    // TODO: show file copy progress
-                    template_child.copy (project_child, FileCopyFlags.NONE);
-                } else {
-                    // remove '.in' suffix if there is one
-                    var project_child_path = (!)project_child.get_path ();
-                    if (!project_child_path.has_suffix (".in"))
-                        warning ("file template %s should have a `.in` suffix", template_child_path_relative);
-                    var renamed_project_child = File.new_for_path (/\.in$/.replace (project_child_path, -1, 0, ""));
-                    template_files[template_child] = renamed_project_child;
-                }
+            if (relative_path == "template.json")
+                continue;   // Don't copy over template.json
+
+            /* Substitute path name */
+            relative_path = Expression.expand_variables (relative_path, variables);
+
+            var project_file = project_dir.resolve_relative_path (relative_path);
+
+            if (file_type == DIRECTORY) {
+                /* Create an empty directory */
+                DirUtils.create_with_parents ((!) project_file.get_path (), 0755);
+                continue;
             }
-        }
 
-        // perform template substitutions
-        foreach (var template_file in template_files.get_keys_as_array ()) {
```

Comment:
It looks like you're performing template substitutions on every file by default now? If so, I think this is a **very bad idea**. That is because assets, binary files, and other project files may contain valid template strings by accident and when contributing new templates this can lead to subtle bugs. Additionally, this could be a performance problem in the future if we have templates with many files or large binary files. The current way--only substituting the files with a `.in` suffix, and specifying these files in the `template.json`--makes it easy to understand how Valdo will transform a template with no surprises.

Reply:
Okay, I understand. It's really bad idea to just treat every file as template. But .in suffix is really useless. May I just bring back `inputs` field without .in suffix?

Reply:
> But .in suffix is really useless.

You're correct that it's redundant, but sometimes redundancy is good when it avoids errors. Anyone checking the templates can easily see which files will be substituted by looking for the `.in` suffix. Or if the template file is copied somewhere else, then there's no need to remember it is a template because it has the `.in` suffix.

I think this comes down to style and tradeoffs: your way is probably fine too, but I don't see it being better than mine. Unless you can convince me otherwise...

Reply:
> Anyone checking the templates can easily see which files will be substituted by looking for the .in suffix.

No. .in suffix is also used by build systems such as Meson for template files. Won't it be better to use some less common suffix, maybe `.template` or something like it?

Reply:
@liferooter maybe `.tmpl`?

Reply:
`.tmpl` is another possibility, but `.in` is used across the board by many programs for anything that's an input to something else. There are even files with `.in.in`. It's also nice to use `.in` because syntax highlighters recognize it.

Reply:
Okay, I'll bring back `.in` suffixes

Reply:
> > But .in suffix is really useless.
> 
> You're correct that it's redundant, but sometimes redundancy is good when it avoids errors. Anyone checking the templates can easily see which files will be substituted by looking for the `.in` suffix. Or if the template file is copied somewhere else, then there's no need to remember it is a template because it has the `.in` suffix.
> 
> I think this comes down to style and tradeoffs: your way is probably fine too, but I don't see it being better than mine. Unless you can convince me otherwise...

Hmm, I think using `.in` suffix makes template creating less convenient and template structure less clear. Also some templates (like GNOME template that I'll finish just after merging this PR because it can't be finished without one fix from it)  can contain files with .in suffix, so this suffix doesn't really shows whether the file is template file. Not-using the suffix allows to simplify template logic and structure. Files are just files, but some of them have variables in its name and content, and all metadata is incapsulated in `template.json`. With suffixes this metadata is smeared on the template structure.

Reply:
Also if template use Meson templates and localization using of `.in` suffixes can cause really weird and non-clear file names  like `${APP_ID}.appdata.xml.in.in.in`

Reply:
@liferooter What are your objections, if any, to having a `.vldo` file extension then?

Reply:
No, I think the best idea to drop suffix at all. Yes, both ways are tradeoffs, but dropping suffix allows to naturally edit and view template reading and writing, allows any editor to correctly understand file type and make template tree more clear. Template files list is for program, not for people.

Reply:
Alright. I think you make good points.

Reply:
So we drops suffixes?

Reply:
@liferooter yes